### PR TITLE
add custom build support

### PR DESCRIPTION
**Pull request by kngwyu**
When we're editing 'build.rs', it's target kind in cargo metadata is 'custom-build', which we can not use as build arguments.
So flycheck tries to run command like `cargo test --custom-build ..` and fails.
To solve this, when cargo metadata returns that current buffer is 'custom-build' target, we have to search other target.

Here's the repository I used for test.
https://gitlab.com/kngwyu/flycheck-rust-buildscript-test

### CODE REVIEW

**Comment by kngwyu** [comment]
added test crates

**Comment by fmdkdd** [comment]
Thanks!  That's a tricky corner case.  

> we have to search other target.

According to the doc:

> The Rust file designated by the build command will be compiled and invoked before anything else is compiled in the package, allowing your Rust code to depend on the built or generated artifacts.

So your solution is: If we are looking at a `build.rs` file (or however it is named), then try to compile any other valid target in order to get errors from the build script.  Correct?

If so, is there a way to call `cargo check` /only/ on the build script?

**Comment by kngwyu** [comment]
> So your solution is: If we are looking at a build.rs file (or however it is named), then try to compile any other valid target in order to get errors from the build script. Correct?

Yes.
> If so, is there a way to call cargo check /only/ on the build script?

I'm not sure, but probably none.

File: flycheck-rust.el [hunk 1]
```diff
@@ -150,7 +150,21 @@ description of the conventional Cargo project layout."
                                             (file-name-directory manifest)))))
             ;; If all else fails, just pick the first target
             (car targets))))
-      (let-alist target (cons (flycheck-rust-normalize-target-kind .kind) .name)))))
+      (let-alist target
```
Comment (fmdkdd) [inline flycheck-rust.el]: Couldn't we just filter out the `custom-build` target from the `targets` list before?  We should fallback on picking the first one without having to iterate through targets again.

Comment (kngwyu) [reply flycheck-rust.el]: Hmm...
I think it'll be more complex if we do so.

Comment (fmdkdd) [reply flycheck-rust.el]: Ah, I see, because we need to make sure to pick up a target *from the same package*, and the relation between packages is lost by `flycheck-rust-get-cargo-targets`.

I'm worried about the way the target from the same package is picked then:
```
(string-prefix-p
  (expand-file-name "src" (file-name-directory file-name))
  .src_path))) 
```
There is no indication that it is, indeed, part of the same package, as returned by `cargo metadata`.

What about doing some pre-processing at the end of `flycheck-rust-get-cargo-targets` to replace `custom-build` targets by another target from the same package?  This would leave `flycheck-rust-find-cargo-target` untouched, and would have the same result, but with the additional guarantee that the `custom-build` replacement comes from the same package.

Comment (kngwyu) [reply flycheck-rust.el]: >There is no indication that it is, indeed, part of the same package, as returned by cargo metadata.

OK.
We have to search targets only from same packages as current buffer.

> What about doing some pre-processing at the end of flycheck-rust-get-cargo-targets to replace custom-build targets by another target from the same package? 

If we do so, how we know a target is from the same package as current `build.rs`?
I think it's natural to process after finding that current target is `custom-build`, because it's a tricky corner case, as you say.

Comment (fmdkdd) [reply flycheck-rust.el]: > If we do so, how we know if a target is from the same package as current build.rs?

Sorry for being unclear.  In the `build-script-test` package, we have two targets, as returned by `cargo metadata`:
```
 "targets": [
        {
          "kind": [ "bin" ],
          "crate_types": [ "bin" ],
          "name": "build-script-test",
          "src_path": ".../build-script-test/src/main.rs"
        },
        {
          "kind": [ "custom-build" ],
          "crate_types": [ "bin" ],
          "name": "build-script-build",
          "src_path": ".../build-script-test/build.rs"
        }
      ],
```
I was suggesting that, before extracting these targets to pass them to `flycheck-rust-find-cargo-target`, we do the following processing:
```
 "targets": [
        {
          "kind": [ "bin" ],
          "crate_types": [ "bin" ],
          "name": "build-script-test",
          "src_path": ".../build-script-test/src/main.rs"
        },
        {
          "kind": [ "bin" ],
          "crate_types": [ "bin" ],
          "name": "build-script-test",
          "src_path": ".../build-script-test/build.rs"
        }
      ],
```
That is, we change the `kind`, and `name` properties of `custom-build` targets to the values of the first target in the same package.  That way, `flycheck-rust-find-cargo-target` will find a matching target for the `build.rs` file, and use the `kind` and `name` values of the first target.

Comment (kngwyu) [reply flycheck-rust.el]: Thanks for explanation.
I really misunderstand your `pre-processing`, and, it looks good.
I'll try it.

Comment (fmdkdd) [reply flycheck-rust.el]: >  honestly now I feel this implementation more complicated than the current one.

I'll trust you on that.  Glad we could discuss alternatives.

File: flycheck-rust.el [hunk 2]
```diff
@@ -134,7 +134,7 @@ description of the conventional Cargo project layout."
             ;; that one
             (seq-find (lambda (target)
                         (let-alist target (string= file-name .src_path)))
-                      targets)
+                      (-flatten-n 1 targets))
```
Comment (fmdkdd) [inline flycheck-rust.el]: I think we should rename `targets` to `packages` in the `-when-let*` above, since that's what they are.  We could then add a `(targets (-flatten-n 1) packages)` binding to that `-when-let*`, to avoid duplication.

File: flycheck-rust.el [hunk 3]
```diff
@@ -145,12 +145,25 @@ description of the conventional Cargo project layout."
...
-                   'cons targets
+                   'cons (-flatten-n 1 targets)
                    (flycheck-rust-dirs-list file-name
                                             (file-name-directory manifest)))))
             ;; If all else fails, just pick the first target
-            (car targets))))
-      (let-alist target (cons (flycheck-rust-normalize-target-kind .kind) .name)))))
+            (car (car targets)))))
+      (let-alist target
+        ;; If target is 'custom-build', we search other target
```
Comment (fmdkdd) [inline flycheck-rust.el]: Please change `we search other target` to `we pick another target from the same package (see GH-62)`.

File: flycheck-rust.el [hunk 4]
```diff
@@ -145,12 +145,25 @@ description of the conventional Cargo project layout."
...
+                 ;; filtering same packages as current buffer
+                 (--filter
+                  (--any?
+                   (let-alist it
+                     (string= file-name .src_path)) it) it)
+                 (--remove
+                  (let-alist it
+                    (string= "custom-build" (car .kind))) (car it))
+                 (let-alist (car it) (cons (car .kind) .name)))
+          (cons (flycheck-rust-normalize-target-kind .kind) .name))))))
```
Comment (fmdkdd) [inline flycheck-rust.el]: I think this can be simplified to:

```emacs-lisp
      ;; If target is 'custom-build', we search other target
      (when (string= "custom-build" (let-alist target (car .kind)))
        (setq target (--> packages
                          ;; filtering same packages as current buffer
                          (--find (--any? (equal target it) it) it)
                          (--find (not (equal target it)) it))))
      (when target
        (let-alist target
          (cons (flycheck-rust-normalize-target-kind .kind) .name))))))
```

This uses `equal` instead of `string=` comparisons to find the right package and target, and it makes sure we normalize the target name in both cases.  It also returns `nil` when there are no other targets in the package containing the `custom-build` package, should that case arrive.

File: tests/test-rust-setup.el [hunk 5]
```diff
@@ -118,4 +121,14 @@
      (expect
       (car (flycheck-rust-find-cargo-target (lib-crate-file "src/lib.rs")))
       :to-equal "lib"))
+
+ (it "'build.rs' to any target in the same workspace member(parent)"
```
Comment (fmdkdd) [inline tests/test-rust-setup.el]: Please add a space before `(parent)`.

File: tests/test-rust-setup.el [hunk 6]
```diff
@@ -118,4 +121,14 @@
...
+     (expect
+      (flycheck-rust-find-cargo-target (build-script-crate-file "build.rs"))
+      :to-equal (cons "bin" "build-script-test")))
+
+ (it "'build.rs' to any target in the same workspace member(child)"
```
Comment (fmdkdd) [inline tests/test-rust-setup.el]: Please add a space before `(child)`.

**Comment by kngwyu** [comment]
Thanks for review.
I modified.

**Comment by fmdkdd** [comment]
Great!  Thank you very much for improving flycheck-rust! :clap:
